### PR TITLE
Reuse Base.ReshapeArray/ReinterpretArray

### PR DESCRIPTION
**Pull request by maleadt**
Currently, `reshape` and `reinterpret` are expected to yield the original array type, complicating that type's definition (it should somehow refcount the underlying buffer, or otherwise support mutating the object). Furthermore, with `view` we sometimes (in the non-contiguous case) need to return a `SubArray` anyhow. So let's just get rid of that behavior, re-use `Base.ReshapedArray` and `ReinterpretArray`.

As a result, downstream packages (i.e. CUDA.jl, AMDGPU.jl, oneAPI.jl) will need to make sure operations that previously only needed to accept a `::GPUArray` now also accept a `ReinterpretArray{<:CuArray}`, and the same for `ReshapeArray`. The reference implementation here does so using a `DenseJLArray` type, for methods that work with (a pointer to) contiguous memory. Next to that, this PR also adds `StridedJLArray` for supporting methods that accept a pointer and a stride (not used here, but interesting for working with CUBLAS in https://github.com/JuliaGPU/CUDA.jl/pull/435), and we already had `WrappedJLArray` for types that we can convert to something GPU-compatible (i.e. for kernels).

TL;DR: remove `GPUArrays.unsafe_reinterpret`, don't implement `Base._reshape`, add a `DenseGPUArray` definition and use that with methods that previously accepted `::GPUArray` and converted it to a pointer. For bonus points: do the same for `view` and always have it return a `SubArray`, introduce a `StridedGPUArray` type and use it for methods that can work with a pointer+stride, and together with the `reshape`/`reinterpret` changes this should allow removing all buffer refcounting from your GPU array type.

cc @jpsamaroo

### CODE REVIEW

**Comment by timholy** [comment]
xref https://github.com/JuliaLang/julia/pull/37559, in case that meddles in your plans here.

**Comment by maleadt** [comment]
Thanks for the link; I think we should be fine if we adapt the Strided/Dense aliases here appropriately. For GPU compatibility, we currently "materialize" wrappers like this when invoking a kernel, so we should be fine.
